src/cbc_sdk/enterprise_edr/threat_intelligence.py

```diff
@@ -509,18 +1139,28 @@ def save_watchlist(self):
         return self
 
     def validate(self):
-        """Validates this report's state.
+        """
+        Validates this report's state.
 
         Raises:
-            InvalidObjectError: If the report's state is invalid
+            InvalidObjectError: If the report's state is invalid.
         """
+        id_substitute = False
+        if not self._info.get('id', None):
+            self._info['id'] = '*'
+            id_substitute = True
         super(Report, self).validate()
+        if id_substitute:
+            del self._info['id']
 
         if self.link and not validators.url(self.link):
             raise InvalidObjectError("link should be a valid URL")
 
         if self.iocs_v2:
             [ioc.validate() for ioc in self._iocs_v2]
+        if self._iocs_v2_need_sync:
```

Comment:
This is a confusing variable name as it doesn't need sync'd instead there are local changes which need pushed. I interpret need_sync as there are new changes on the CBC backend which need pulled.

Reply:
Consider using a name schema like dirty as we use that already in https://github.com/carbonblack/carbon-black-cloud-sdk-python/blob/develop/src/cbc_sdk/base.py#L376

Reply:
need_sync is kind of different from that.  The Report object actually contains _two_ representations of the current list of reports, the JSON representation in the info array and the list of Report objects.  That flag gets set to True when you add an object to the array, it records that the JSON needs to be rebuilt.

Reply:
It also gets set when you _remove_ an object from the array, too. Either way.

Reply:
Looking over all the places we use `_iocs_v2_need_sync`. Why wait till the validate Reports to check the IOCs and push that into `_info` it seems like we should be doing that were we replace or add reports. How does this work if I update an IOC directly? What is the design for updating an existing IOC?

Reply:
Because we don't want to rebuild that big list of dicts every time we do something to an IOC in the report. It wastes time and memory. We want to encourage the use of the objects whereever possible.  At the same time, we _also_ don't want to have to rebuild that big list of dicts if the IOCs haven't been touched, which, again, wastes time and memory.

Reply:
Thanks for the explanation. I'm stuck on the name as it's not intuitive to me exactly what it represents though I do understand it better now. A small comment here https://github.com/carbonblack/carbon-black-cloud-sdk-python/blob/8df656cd6a540e4a181c899b6cdf5691569224b3/src/cbc_sdk/enterprise_edr/threat_intelligence.py#L965 will be enough to satisfy my confusion

Reply:
I added the comment and also changed it to `_iocs_v2_need_rebuild`, as that's a better word for what it does (flags that we need to rebuild the `ioc_v2` element of `_info` based on what's in the `iocs_v2` array).

Reply:
Thanks! This is much clearer to me now